### ISSUE TITLE
[question] Porting 1.x imports() to 2.0

### ISSUE THREAD

**Issue by db4** [conan-io/conan#12393]
I'm following [Conan migration guide to 2.0](https://docs.conan.io/en/latest/conan_v2.html#conan-migration-guide-to-2-0). It contains the following recommended pattern that I expected to work everywhere:
```python
from conan.tools.files import copy

def generate(self):
    for dep in self.dependencies.values():
        copy(self, "*.dylib", dep.cpp_info.libdirs[0], self.build_folder)
        copy(self, "*.dll", dep.cpp_info.libdirs[0], self.build_folder)
```
Unfortunately `self.build_folder` is set only in 2.0 and is None in 1.x so for 1.x I have to use `self.build_folder` that in turn doesn't exist in 2.0. The following code works:

```python
    def generate(self):
        cpp_info = self.dependencies["some_dependency"].cpp_info
        dst = self.build_folder if self.build_folder else self.folders.build
        rootpath = os.path.dirname(cpp_info.bindirs[0])
        copy(self, "*.dat", rootpath, dst=dst)
```
but I would prefer something working out of the box. BTW, why `rootpath` attribute is excluded from the new dependency model? 

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by czoido** [comment]
Hi @db4,

I think that maybe what's happening here is that the recipe does not have declared the [layout()](https://docs.conan.io/en/latest/developing_packages/package_layout.html). Does your Conan 1.X recipe have the `layout()` method declared? Something like:

```
def layout(self):
    cmake_layout(self)
```

Could that be the issue? Hope this helps

**Comment by db4** [comment]
Hi @czoido,
No, layout() is there, exactly the same as yours. Conan 1.53.0, 2.0.0-beta4

**Comment by czoido** [comment]
Hi @db4,

For me, doing both `conan create .` and `conan install .` on this _conanfile.py_ that checks that `self.build_folder` exists is working for both conan 1.53 and conan 2.0.0-beta4. Could you provide an example to reproduce the issue?

```
from conan import ConanFile
from conan.tools.cmake import cmake_layout

class libConan(ConanFile):
    name = "lib"
    version = "1.0"
    settings = "os", "compiler", "build_type", "arch"
    
    def layout(self):
        cmake_layout(self)

    def generate(self):
        assert self.build_folder is not None
        print("*"*10, self.build_folder)
```

Thanks.

**Comment by db4** [comment]
Hi @czoido,
I'm doing it in `test_package`. Maybe that's the reason?

**Comment by db4** [comment]
Just added your generate() method to zlib/1.2.12 test_package from CCI:
```
conan create . zlib/1.2.12@_/_
...
zlib/1.2.12 (test package): Preset 'default' added to CMakePresets.json. Invoke it manually using 'cmake --preset default'
zlib/1.2.12 (test package): If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G "Visual Studio 15 2017" -DCMAKE_TOOLCHAIN_FILE=C:\Users\dbely\conan\conan-center-index\recipes\zlib\all\test_package\build\generators\conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW'
zlib/1.2.12 (test package): Calling generate()
ERROR: zlib/1.2.12 (test package): Error in generate() method, line 16
        assert self.build_folder is not None
        AssertionError:
```

**Comment by czoido** [comment]
Hi @db4,
Yes, calling it in the test_package is the reason, may I ask what's the use case and why do you need to call `copy` to `self.build_folder` in the test_package?
Thanks

**Comment by db4** [comment]
I need to import some additional data files from a package to run tests: they expect to find them in ../data directory (relative to bin)

**Comment by czoido** [comment]
Hi @db4,
The _test_package_ should only be used to test that the Conan package was correctly created and works, not for running your tests over the library or application, if you want to run tests over that, you should do it in the `build()` method of the package, there's an example in the Conan 2.0 documentation here: https://docs.conan.io/en/2.0/tutorial/creating_packages/build_packages.html#build-and-run-tests-for-your-project
Hope this helps.

**Comment by db4** [comment]
@czoido here is my situation (oversimplified):
```
module1
└─── bin
│    │ module1.dll
└─── data
     │ module1.dat

module2
└─── bin
│    │ module2.dll
└─── data
     │ module2.dat

application
└─── bin
│    │ app.exe
│    │ module1.dll
│    │ module2.dll
└─── data
     | module1.dat
     | module2.dat
```
module2 depends on module1.  Just to load module2 and create a root C++ object it exports, its dependencies (module1.dll and module2.dat) have to be placed in `.` and `../data` folders respectively (relative to module2.dll). So to test module2 I have to import module1's files in test_package. I always did it in Conan 1.x via import(), I'll do the same in Conan 2.0, I just prefer to do it the uniform way. But if it's not possible - ok, I will work around pitfalls.

**Comment by db4** [comment]
@czoido the question is still open: is it so hard to make `self.build_folder` work in `test_package` (Conan 1.x)? Why? After all, isn't it a (special) consumer recipe where all the usual features are expected to be functional?

**Comment by czoido** [comment]
Hi @db4,
Sorry for the delay in answering back. Yes, after seeing your use case, I think it makes sense that this is available in the test_package as well. I'll mark this for the next release to have a look.
Thanks!

**Comment by memsharded** [comment]
Fixed by https://github.com/conan-io/conan/pull/12443 to be released in 1.54
